[DataPipe] implementing demux()

NivekT opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #64215
* __->__ #63650
* #63649

Differential Revision: [D30493944](https://our.internmc.facebook.com/intern/diff/D30493944)

cc @VitalyFedyunin @ejguan

NivekT committed:
implementing demux()

[ghstack-poisoned]

VitalyFedyunin commented on torch/utils/data/datapipes/iter/combining.py:
```diff
@@ -137,24 +135,70 @@ def __iter__(self):
+            value = self.main_datapipe.__next__()
+            classification = self.classifier_fn(value)
+            if classification > self.num_instances:
+                raise ValueError(f"Output of the classification fn should be between 0 and {self.num_instances - 1}. " +
```
Add test

NivekT replied:
Added.

VitalyFedyunin commented on torch/utils/data/datapipes/iter/combining.py:
```diff
@@ -137,24 +135,70 @@ def __iter__(self):
+        self.num_instances = num_instances
+        self.buffer_size = buffer_size
+        self.child_buffers: List[Deque[T_co]] = [deque() for _ in range(num_instances)]
+        self.current_buffer_total = 0
```
nit: s/current_buffer_total/total_buffers_size/g

NivekT replied:
I renamed the variables to
```
self.max_buffer_size = buffer_size
self.current_buffer_usage = 0
```

VitalyFedyunin commented on torch/utils/data/datapipes/iter/combining.py:
```diff
@@ -137,24 +135,70 @@ def __iter__(self):
+
+    def __init__(self, datapipe: IterDataPipe[T_co], num_instances: int,
+                 classifier_fn: Callable[[T_co], int], buffer_size: int):
+        self.main_datapipe = iter(datapipe)
```
drop `iter` please

NivekT replied:
As discussed in the other PR, I need `iter` for the subsequent `__next__` call.

VitalyFedyunin reviewed (commented):
Couple changes required

NivekT committed:
Update on "[DataPipe] implementing demux()"




[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] implementing demux()"




[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] implementing demux()"




[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30493944).

NivekT committed:
Update on "[DataPipe] implementing demux()"


Differential Revision: [D30493944](https://our.internmc.facebook.com/intern/diff/D30493944)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30493944).

NivekT committed:
Update on "[DataPipe] implementing demux()"


Differential Revision: [D30493944](https://our.internmc.facebook.com/intern/diff/D30493944)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30493944).

ejguan reviewed (commented):
A few comments similar to `Fork`

NivekT committed:
Update on "[DataPipe] implementing demux()"


Differential Revision: [D30493944](https://our.internmc.facebook.com/intern/diff/D30493944)

[ghstack-poisoned]

ejguan commented:
Can you please also add `Demux` and `Mux` to https://github.com/pytorch/pytorch/blob/master/torch/utils/data/datapipes/iter/__init__.py to expose these APIs for users?

NivekT committed:
Update on "[DataPipe] implementing demux()"


Differential Revision: [D30493944](https://our.internmc.facebook.com/intern/diff/D30493944)

[ghstack-poisoned]

ejguan commented:
Forward feedback from vision domain for `demux`. Can we provide a flag of `drop_none`? When the result of `classification_fn` is `None`, `Demux` can ignore the data without putting it into buffer.

NivekT committed:
Update on "[DataPipe] implementing demux()"


Differential Revision: [D30493944](https://our.internmc.facebook.com/intern/diff/D30493944)

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] implementing demux()"


Differential Revision: [D30493944](https://our.internmc.facebook.com/intern/diff/D30493944)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30493944).

ejguan reviewed (approved):
LGTM

NivekT committed:
Update on "[DataPipe] implementing demux()"


Differential Revision: [D30493944](https://our.internmc.facebook.com/intern/diff/D30493944)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30493944).

NivekT committed:
Update on "[DataPipe] implementing demux()"


Differential Revision: [D30493944](https://our.internmc.facebook.com/intern/diff/D30493944)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30493944).
